syntax = "proto3";
import "google/protobuf/empty.proto";

option java_multiple_files = true;
option objc_class_prefix = "HLW";

package controlcommands;

service ControlCommands {
  rpc ExecuteNetworkTest(LoopRange) returns (google.protobuf.Empty) {}
  rpc StartContainer(ContainerConfig) returns (google.protobuf.Empty) {}
  rpc UpdateDownstream(ContainerLink) returns (google.protobuf.Empty) {}
  rpc SyncDatasource(ContainerLink) returns (google.protobuf.Empty) {}
  rpc UpdateBatchSize(ContainerInts) returns (google.protobuf.Empty) {}
  rpc UpdateResolution(ContainerInts) returns (google.protobuf.Empty) {}
  rpc UpdateTimeKeeping(TimeKeeping) returns (google.protobuf.Empty) {}
  rpc StopContainer(ContainerSignal) returns (google.protobuf.Empty) {}
}

message ContainerConfig {
  string name = 1;
  string json_config = 2;
  string executable = 3;
  int32 device = 4;
  int32 control_port = 5;
}

message TimeKeeping {
  string name = 1;
  int32 slo = 2;
  int32 cont_slo = 3;
  uint64 time_budget = 4;
  uint64 start_time = 5;
  uint64 end_time = 6;
  uint64 local_duty_cycle = 7;
  uint64 cycle_start_time = 8;
}

message ContainerLink {
  int32 mode = 1;
  string name = 2;
  string downstream_name = 3;
  string ip = 4;
  int32 port = 5;
}

message ContainerSignal {
  string name = 1;
  bool forced = 2;
}

message ContainerInts {
  string name = 1;
  repeated int32 value = 2;
}

message LoopRange {
  int32 min = 1;
  int32 max = 2;
  int32 repetitions = 3;
}